node/src/components/chain_synchronizer/operations.rs

```diff
@@ -334,17 +334,27 @@ where
         .ok_or_else(|| Error::HeightOverflow {
             parent: Box::new(parent_header.clone()),
         })?;
-    let mut peers = effect_builder.get_fully_connected_peers().await.into_iter();
+
+    // Ensure we have at least one peer to which we're bidirectionally connected before trying to
+    // fetch the data.
+    let mut peers = vec![];
+    for _ in 0..config.max_retries_while_not_connected() {
+        peers = effect_builder.get_fully_connected_peers().await;
+        if !peers.is_empty() {
+            break;
+        }
+        tokio::time::sleep(config.retry_interval()).await;
+    }
```

Comment:
So much code just for these few lines? :)

Reply:
Yeah, I know :(

How I got here:
* I wanted to wait in the `fetch_and_store_next` function until we have at least one good connection... but not indefinitely in case we never get one (probably waiting for more than one would be good, but adding another config option for that too seemed a step too far - I could be persuaded otherwise)
* we need a timeout for that - it should be configurable, and based upon whatever is reasonable to wait for a good connection
* that shouldn't be a separate config file entry - it should be based on a couple of iterations of gossiping our address, so we need a separate config struct just for the chain sync
* we want to eventually remove the chainspec from the chain sync (there's a ticket for that) so this helps lay the groundwork
* the sleep duration should be configurable too

So while some of these changes aren't strictly required, I feel they're a step in the right direction.